### ISSUE TITLE
Where is the Option for Server Workspaces

### ISSUE THREAD

**Issue by fifthgearonline** [microsoft/azure-pipelines-agent#2586]
## Have you tried troubleshooting?
I have not found any way to use Sever workspaces since the introduction of the 2.x agents with TFS 2017. Some of my builds have over 250,000 files and server workspaces are absolutely mandatory for performance reasons. 

## Agent Version and Platform
2.144.2
Windows

## Azure DevOps Type and Version
Azure Devops 2019 Update 1

## What's not working?
According to release V2.105.2  There was to be a web option added for using server workspaces
* "tfsvc uses local workspaces (like 1.x) - will add web UI option for server workspaces."
https://github.com/microsoft/azure-pipelines-agent/releases/tag/v2.105.2

There is one problem with that statement. The 1.x agents used the collections default workspace type, they did NOT force local workspaces. 

[https://social.msdn.microsoft.com/Forums/en-US/8d1181f3-ed9c-46ca-847d-9ea6946d75c5/tfs-2017-build-agent-only-uses-local-workspaces-instead-of-server?forum=tfsbuild
](https://social.msdn.microsoft.com/Forums/en-US/8d1181f3-ed9c-46ca-847d-9ea6946d75c5/tfs-2017-build-agent-only-uses-local-workspaces-instead-of-server?forum=tfsbuild
)

* Why were the agents changed to no longer respect the collections default workspace type?
* Has that server workspace option been added?
* If so when was it added, and where is the option located at?
* What is the issue number for that code change that added the web-ui option for server workspaces. 

According to microsofts own documentation Server workspaces are recommended for builds with over 100,000 item. 

[https://docs.microsoft.com/en-us/azure/devops/repos/tfvc/decide-between-using-local-server-workspace?view=azure-devops](https://docs.microsoft.com/en-us/azure/devops/repos/tfvc/decide-between-using-local-server-workspace?view=azure-devops)

```
Important:One drawback to using a local workspace is that performance degrades as the number of items increases. See the next section for details.
When might I need to use a server workspace?
When you use a server workspace, Visual Studio keeps only one copy of each file. This can significantly reduce disk space usage and improve performance when you have a lot of items. We recommend that you use a server workspace if:
Your workspace contains more than 100,000 items.
```

## Agent and Worker's Diagnostic Logs

This was tested with Azure Devops Server Update 1

When running a build with a large amount of files, a warning is given at the end of the "checkout" step that states

```
"2019-11-05T15:55:52.4792639Z TF401190: The local workspace ws_1_153;Project Collection Build Service (Collection) has 250000 items in it, which exceeds the recommended limit of 100000 items. To improve performance, either reduce the number of items in the workspace, or convert the workspace to a server workspace."
```

I was able to manually flip that workspace from Local, to Server as it suggests. However, the next time the build runs, it forces the workspaces BACK to a local workspace. That defeats the entire purpose of setting it to a server workspace. The V1 agents actually respected the collection's default workspace type, the V2 agents do not and force a local workspace, even with microsoft's own warning messages advise against that.

Later i queued a new build and received this at the beginning of my build. It as you can see its deleting the workspace and creating a new local workspace. Server workspaces need to be an option for those of us working with large enterprise class projects. 

I generalized the paths in the log below so the exact mappings are not going to line up, as that's not the problem.

```
2019-11-05T16:18:47.9790191Z Querying workspace information.
2019-11-05T16:18:48.5058384Z ##[command]tf vc workspace /delete ws_1_153%3Ba9d19f6f-1356-4c52-9e99-3c98f15d354f /collection:https://server:8080/tfs/project/ /loginType:OAuth /login:.,*** /noprompt
2019-11-05T16:19:46.5467493Z ##[command]tf vc workspace /new /location:local /permission:Public ws_1_153 /collection:https://server:8080/tfs/project/ /loginType:OAuth /login:.,*** /noprompt
2019-11-05T16:19:47.1634398Z ##[command]tf vc workfold /unmap /workspace:ws_1_153 $/ /collection:https://server:8080/tfs/project/ /loginType:OAuth /login:.,*** /noprompt
2019-11-05T16:19:47.7136685Z ##[command]tf vc workfold /map /workspace:ws_1_153 $/Project/Current/Tools/Build C:\Builds\1\s\Project\Tools\Build /collection:https://server:8080/tfs/project/ /loginType:OAuth /login:.,*** /noprompt
2019-11-05T16:19:48.2342290Z ##[command]tf vc workfold /map /workspace:ws_1_153 $/Project/Current C:\Builds\1\s\Project\Current /collection:https://server:8080/tfs/project/ /loginType:OAuth /login:.,*** /noprompt
2019-11-05T16:19:48.7800983Z ##[command]tf vc workfold /map /workspace:ws_1_153 $/Project2/Current C:\Builds\1\s\Project2\Current /collection:https://server:8080/tfs/project/ /loginType:OAuth /login:.,*** /noprompt
2019-11-05T16:19:49.3279795Z ##[command]tf vc workfold /cloak /workspace:ws_1_153 $/Project/Current/ReleaseInfo /collection:https://server:8080/tfs/project/ /loginType:OAuth /login:.,*** /noprompt
2019-11-05T16:19:49.8171539Z ##[command]tf vc workfold /cloak /workspace:ws_1_153 $/Project2/Current/ReleaseInfo /collection:https://server:8080/tfs/project/ /loginType:OAuth /login:.,*** /noprompt
2019-11-05T16:19:50.3495721Z ##[command]tf vc get /version:138101 /recursive /overwrite C:\Builds\1\s /loginType:OAuth /login:.,*** /noprompt
```

**Comment by fifthgearonline** [comment]
@ericsciple can you comment as to why the functionality of server workspaces was removed compared to the V1 agents? Where is the ui option that was supposed to be added? 

[](https://github.com/microsoft/azure-pipelines-agent/commit/34535bd1e918353a346eb3c79359e40c529fed65)

**Comment by jtpetty** [comment]
@vtbassmatt - Can you comment on this? Do you know the story here?

**Comment by vtbassmatt** [comment]
The move from v1 to v2 of the agent happened 3+ years ago and predates everyone now on the team. I'll see if I can get some history on the decision.

@fifthgearonline were you using the 1.x agent until very recently? I'm surprised we haven't gotten this feedback many times prior to now. My fear is that you might have, for instance, just upgraded from an older TFS instance and therefore are the "tip of the spear" in a wave of coming complaints.

**Comment by fifthgearonline** [comment]
We have actually been using the 2.x agents for several years now, and have just been dealing with the issue. So all of our robocopy commands have xf commands to ignore all local workspace file types. We had to upgrade the agents to 2.x because when 1.x agents are used on 2017 or newer servers, it downloads all source code every time. This was unusable for our CI builds.

**Comment by vtbassmatt** [comment]
Thanks. That at least means we shouldn't expect a tidal wave of new feedback related to, say, shipping Server 2019.1.1. Stay tuned, I'm getting some more info on this.

**Comment by anmason** [comment]
An option to specify server workspaces is added here: https://github.com/microsoft/azure-pipelines-agent/pull/2663

**Comment by jtpetty** [comment]
This was included in agent 2.164.6
